### PR TITLE
Unify the replication calculation between query and push.

### LINKED ISSUE #279: Unify the replication logic into a single place
Probably want some interfaces and alternative implementations, although I don't particularly think we'll ever use any thing other than quorum...

### PR DESCRIPTION
**Pull request by tomwilkie**
Part of #604, fixes #279

The logic for doing replication was subtlety different on the query side and on the push side - push used the number of ingesters returned to calculate quorum, and query used the replication factor.  Using the replication factor to calculate quorum will occasionally give the wrong answer when ingesters are joining and leaving.  Whats more, there were a bunch of checks on the push path that weren't on the query path.

This PR unifies the logic between these two paths, effectively adding a bunch of checks on the query path and dealing with the case where you have less ingesters than the replication factor (by returning an appropriate error).

### CODE REVIEW

**Comment by tomwilkie** [comment]
(found a bug with this, don't merge just yet)

**Comment by tomwilkie** [comment]
After some more work, I realised all this replication logic really belongs to the ring, as does the replication factor itself.  This reduces coupling quite a lot IMO, and the ring interface now becomes "tell me who to talk to and how many errors to tolerate", with all the liveness filtering and quorum calculations happening in the ring.

Going to deploy to our dev env and see how this works.

**Comment by tomwilkie** [comment]
Have put this through a couple of rolling upgrades and it works, I'd say its ready for review.  @bboreham fancy this one?

File: pkg/ring/ring.go [hunk 1]
```diff
@@ -195,47 +212,61 @@ func (r *Ring) getInternal(key uint32, n int, op Operation) ([]*IngesterDesc, er
...
-		// not want to Write to them because they are about to go away, but we do
-		// want to write the extra replica somewhere.  So we increase the size of the
-		// set of replicas for the key.  This means we have to also increase the
+		// We do not want to Write to Ingesters that are not ACTIVE, because they are
+		// about to go away, but we do want to write the extra replica somewhere.
+		// So we increase the size of the set of replicas for the key.
+		// This means we have to also increase the
 		// size of the replica set for read, but we can read from Leaving ingesters,
 		// so don't skip it in this case.
+		// NB ingester will be filterer later (by distributor/replication_strategy).
```
Comment (bboreham) [inline pkg/ring/ring.go]: spelling "filterer"

File: pkg/ring/ring.go [hunk 2]
```diff
@@ -195,47 +212,61 @@ func (r *Ring) getInternal(key uint32, n int, op Operation) ([]*IngesterDesc, er
 		distinctHosts[token.Ingester] = struct{}{}
 		ingester := r.ringDesc.Ingesters[token.Ingester]
 
-		// Ingesters that are not ACTIVE do not count to the replication limit. We do
-		// not want to Write to them because they are about to go away, but we do
-		// want to write the extra replica somewhere.  So we increase the size of the
-		// set of replicas for the key.  This means we have to also increase the
+		// We do not want to Write to Ingesters that are not ACTIVE, because they are
+		// about to go away, but we do want to write the extra replica somewhere.
```
Comment (bboreham) [inline pkg/ring/ring.go]: Nit: could be pending or joining as well as leaving, so "about to go away" is not a full explanation.

File: pkg/distributor/distributor_test.go [hunk 3]
```diff
@@ -24,24 +24,33 @@ type mockRing struct {
 	heartbeatTimeout time.Duration
```
Comment (bboreham) [inline pkg/distributor/distributor_test.go]: Do we still need this?

File: pkg/ring/replication_strategy.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	"fmt"
+	"time"
+)
+
+func (r *Ring) replicationStrategy(ingesters []*IngesterDesc) (
+	liveIngesters []*IngesterDesc, maxFailure int, err error,
+) {
+	// We need a response from a quorum of ingesters, which is n/2 + 1.  In the
+	// case of a node joining/leaving, the actual replica set might be bigger
+	// than the replication factor, so we need to account for this.
```
Comment (bboreham) [inline pkg/ring/replication_strategy.go]: I don't think it is clear what "account for this" means or refers to.

File: pkg/ring/replication_strategy.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package ring
+
+import (
+	"fmt"
+	"time"
+)
+
+func (r *Ring) replicationStrategy(ingesters []*IngesterDesc) (
```
Comment (bboreham) [inline pkg/ring/replication_strategy.go]: please explain what the overall purpose of this function is, and what the `ingesters` parameter should contain

**Comment by tomwilkie** [comment]
Thanks for the feedback; this is quite a complicated and important aspect of Cortex so a second brain is appreciated.

I've kept `replicationStrategy` as a separate function mainly for testing.

**Comment by bboreham** [comment]
AAMOI does this make it easier to do https://github.com/weaveworks/cortex/issues/467#issuecomment-319941721 ?

**Comment by tomwilkie** [comment]
Yes!  We can reevaluate writing out the extra replica when we hit a non-active ingester, given out transfer protocol. 

I'd prefer to fix up the transfers so they do the right thing though - instead of streaming of the new joining ingester, they should really calculate where their data should like once their tokens are gone, and send it there (similarly, joining nodes should 'steal' data from nodes based on the tokens they pick).  This will enable us to much more easily scale up and down, and in this word the current logic in the ring is correct.  But thats more work - I'll get to it this year I suspect.

**Comment by jml** [comment]
@bboreham is not available this week. I'm not likely to get to this before then (although will try).

File: pkg/ring/ring.go [hunk 6]
```diff
@@ -150,38 +164,41 @@ func (r *Ring) loop(ctx context.Context) {
...
 	defer r.mtx.RUnlock()
-	return r.getInternal(key, n, op)
+	return r.getInternal(key, op)
 }
 
-// BatchGet returns n (or more) ingesters which form the replicas for the given key.
-// The order of the result matches the order of the input.
-func (r *Ring) BatchGet(keys []uint32, n int, op Operation) ([][]*IngesterDesc, error) {
+// BatchGet returns ReplicationFactor (or more) ingesters which form the replicas
+// for the given key. The order of the result matches the order of the input.
```
Comment (bboreham) [inline pkg/ring/ring.go]: key _s_

File: pkg/ring/ring.go [hunk 7]
```diff
@@ -84,6 +94,10 @@ type Ring struct {
 
 // New creates a new Ring
 func New(cfg Config) (*Ring, error) {
+	if 0 > cfg.ReplicationFactor {
+		return nil, fmt.Errorf("ReplicationFactor must be greater than zero: %d", cfg.ReplicationFactor)
```
Comment (bboreham) [inline pkg/ring/ring.go]: Error message does not match test.  Suspect you meant `0 >=`

File: pkg/ring/replication_strategy.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+			minSuccess, len(liveIngesters))
+		return
+	}
+
+	return
+}
+
+// IsHealthy checks whether an ingester appears to be alive and heartbeating
+func (r *Ring) IsHealthy(ingester *IngesterDesc) bool {
+	if ingester.State != ACTIVE {
```
Comment (bboreham) [inline pkg/ring/replication_strategy.go]: Looks like this means you no longer read from `LEAVING` ingesters - was that intended?

File: pkg/ring/ring.go [hunk 9]
```diff
@@ -195,47 +212,60 @@ func (r *Ring) getInternal(key uint32, n int, op Operation) ([]*IngesterDesc, er
...
-		// Ingesters that are not ACTIVE do not count to the replication limit. We do
-		// not want to Write to them because they are about to go away, but we do
-		// want to write the extra replica somewhere.  So we increase the size of the
-		// set of replicas for the key.  This means we have to also increase the
+		// We do not want to Write to Ingesters that are not ACTIVE, but we do want
+		// to write the extra replica somewhere.  So we increase the size of the set
+		// of replicas for the key. This means we have to also increase the
 		// size of the replica set for read, but we can read from Leaving ingesters,
 		// so don't skip it in this case.
+		// NB dead ingester will be filtered later (by replication_strategy.go).
```
Comment (bboreham) [inline pkg/ring/ring.go]: dead and not-`ACTIVE` ingesters are filtered by `IsHealthy()`, which doesn't really match the logic here.  Or I am missing the subtlety.

File: pkg/ring/replication_strategy.go (hunk 8 above)
Comment (tomwilkie) [reply pkg/ring/replication_strategy.go]: It wasn't, good catch.

**Comment by tomwilkie** [comment]
Thanks for the review, I've reintroduced the old behaviour.

File: pkg/ring/ring.go [hunk 10]
```diff
@@ -293,7 +323,7 @@ func (r *Ring) Collect(ch chan<- prometheus.Metric) {
 		LEAVING.String(): 0,
 	}
 	for _, ingester := range r.ringDesc.Ingesters {
-		if !r.IsHealthy(ingester) {
+		if !r.IsHealthy(ingester, Write) {
```
Comment (bboreham) [inline pkg/ring/ring.go]: Now you're never going to report LEAVING ingesters

File: pkg/ring/http.go [hunk 11]
```diff
@@ -102,7 +102,7 @@ func (r *Ring) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		ing := r.ringDesc.Ingesters[id]
 		timestamp := time.Unix(ing.Timestamp, 0)
 		state := ing.State.String()
-		if !r.IsHealthy(ing) {
+		if !r.IsHealthy(ing, Write) {
```
Comment (bboreham) [inline pkg/ring/http.go]: probably don't want this to report unhealthy on LEAVING either

**Comment by bboreham** [comment]
@tomwilkie I think it's close; do you want to rebase and consider the reporting aspects?

**Comment by bboreham** [comment]
@tomwilkie I have rebased and added a suggested fix at https://github.com/weaveworks/cortex/compare/604-unify-replication-calcs; PTAL

**Comment by tomwilkie** [comment]
You're branch looks good Bryan; I've pull it in here.  Thanks for doing that.

**Pull request merged**